c10/core/TensorImpl.cpp

```diff
@@ -225,25 +227,27 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 AutogradMetaInterface::~AutogradMetaInterface() {}
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
-  TORCH_INTERNAL_ASSERT(autograd_meta(), "set_requires_grad is not implemented for Tensor");
-  autograd_meta()->set_requires_grad(requires_grad, this);
+  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
+  if (!requires_grad && !autograd_meta_) return;
+  if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
+  autograd_meta_->set_requires_grad(requires_grad, this);
```

Comment:
Discussed with @ezyang offline, and it seems that we have three possible states of a Variable after this PR:
1) Variable with `autograd_meta_ == nullptr`
2) Variable with `autograd_meta_` default constructed
3) Variable with `autograd_meta_` fully populated

I think it would be all good if at the end of the merge we only have 1) and 3), which corresponds to `requires_grad` being false / true, respectively. It would also be awesome to have a comment about the intermediate state :D